File: src/core/worktree.ts
import { mkdir, mkdtemp, rm } from "node:fs/promises";
import { tmpdir } from "node:os";
import { dirname, join } from "node:path";
import { GitRepository } from "./git.js";

export interface TemporaryWorktree {
  path: string;
  repo: GitRepository;
  branchName?: string;
  cleanup(): Promise<void>;
}

export interface CreateWorktreeOptions {
  prefix?: string;
  path?: string;
  branchName?: string;
}

export class WorktreeManager {
  async createFromHead(baseRepo: GitRepository, options: CreateWorktreeOptions = {}): Promise<TemporaryWorktree> {
    const prefix = options.prefix ?? "detdoc-worktree-";
    const container = options.path ? undefined : await mkdtemp(join(tmpdir(), prefix));
    const path = options.path ?? join(container!, "worktree");
    const head = await baseRepo.headCommit();

    if (options.path) await mkdir(dirname(path), { recursive: true });

    if (options.branchName) {
      await baseRepo.git(["worktree", "add", "-b", options.branchName, path, head]);
    } else {
      await baseRepo.git(["worktree", "add", "--detach", path, head]);
    }

    const repo = new GitRepository(path);
    return {
      path,
      repo,
      branchName: options.branchName,
      cleanup: async () => {
        await baseRepo.git(["worktree", "remove", "--force", path]).catch(async () => {
          await rm(path, { recursive: true, force: true });
        });
        if (options.branchName) {
          await baseRepo.git(["branch", "-D", options.branchName]).catch(() => undefined);
        }
        if (container) await rm(container, { recursive: true, force: true });
      },
    };
  }
}

File: src-tauri/src/detdoc/worktree.rs
use std::fs;
use std::path::PathBuf;

use super::error::DetDocResult;
use super::git::GitRepository;

#[derive(Debug, Clone)]
pub struct WorktreeHandle {
    pub path: PathBuf,
    pub branch_name: String,
}

pub struct WorktreeManager;

impl WorktreeManager {
    pub fn new() -> Self { Self }

    pub fn create_from_head(&self, repo: &GitRepository, run_id: &str) -> DetDocResult<WorktreeHandle> {
        let path = repo.cwd.join(".worktrees").join(run_id);
        fs::create_dir_all(repo.cwd.join(".worktrees")).map_err(|error| super::error::DetDocError::new("WORKTREE_DIR_FAILED", error.to_string()))?;
        let path_string = path.to_string_lossy().to_string();
        let base = repo.head()?;
        repo.git(&["worktree", "add", "-b", run_id, &path_string, &base])?;
        Ok(WorktreeHandle { path, branch_name: run_id.to_string() })
    }

    pub fn cleanup(&self, repo: &GitRepository, handle: &WorktreeHandle) -> DetDocResult<()> {
        let path_string = handle.path.to_string_lossy().to_string();
        if handle.path.exists() {
            repo.git(&["worktree", "remove", "--force", &path_string])?;
        }
        repo.git(&["branch", "-D", &handle.branch_name])?;
        Ok(())
    }
}
